packages/xo-server-sdn-controller/src/protocol/openflow-channel.js

```diff
@@ -0,0 +1,393 @@
+            type,
+            // data: openflow.unpack(data),
+          })
+        }
+        break
+      case ofProtocol.type.echoRequest:
+        this._sendPacket(
+          this._syncMessage(ofProtocol.type.echoReply, message.header.xid)
+        )
+        break
+      case ofProtocol.type.packetIn:
+        log.debug('PACKET_IN')
+        break
+      case ofProtocol.type.featuresReply:
+        {
+          const { datapath_id: dpid, capabilities, ports } = message
+          log.debug('FEATURES_REPLY', { dpid, capabilities, ports })
+          this._sendPacket(
+            this._syncMessage(
+              ofProtocol.type.getConfigRequest,
+              message.header.xid
+            )
+          )
+        }
+        break
+      case ofProtocol.type.getConfigReply:
+        {
+          const { flags } = message
+          log.debug('CONFIG_REPLY', { flags })
+          this.emit('ofConnected')
+        }
+        break
+      case ofProtocol.type.portStatus:
+        log.debug('PORT_STATUS')
+        break
+      case ofProtocol.type.flowRemoved:
+        log.debug('FLOW_REMOVED')
+        break
+      default:
+        log.error('Unknown OpenFlow type', { ofType })
+        break
+    }
+  }
+
+  _addFlow(match, instructions) {
+    const packet = this._flowModMessage(
+      ofProtocol.flowModCommand.add,
+      match,
+      instructions
+    )
+    this._sendPacket(packet)
+  }
+
+  _removeFlows(match) {
+    const packet = this._flowModMessage(ofProtocol.flowModCommand.delete, match)
+    this._sendPacket(packet)
+  }
+
+  // ---------------------------------------------------------------------------
+
+  _syncMessage(type, xid = 1) {
+    return {
+      header: {
+        version,
+        type,
+        xid,
+      },
+    }
+  }
+
+  _flowModMessage(command, match, instructions = []) {
+    // TODO: Do not use default priority?
+    return {
+      ...this._syncMessage(ofProtocol.type.flowMod),
+      command,
+      flags: ofProtocol.flowModFlags.sendFlowRem,
+      match,
+      instructions,
+    }
+  }
+
+  // ---------------------------------------------------------------------------
+
+  _sendPacket(packet) {
+    const buf = openflow.pack(packet)
+    try {
+      this._socket.write(buf)
+    } catch (error) {
+      log.error('Error while writing into socket', {
+        error,
+        host: this.host.name_label,
+      })
+    }
+  }
+
+  // ---------------------------------------------------------------------------
+
+  async _parseMessages() {
+    for await (const msg of parse(this._socket)) {
+      if (msg.header !== undefined) {
+        this._processMessage(msg)
+      } else {
+        log.error('Error: Message is unparseable', { msg })
+      }
+    }
+  }
+
+  async _connect() {
+    if (this._socket !== undefined) {
+      return
+    }
+
+    this._socket = await this._tlsHelper.connect(
+      this.host.address,
+      OPENFLOW_PORT
+    )
+
+    const deleteSocket = () => delete this._socket
```

Comment:
FYI, for performance concerns, it's best not to add/remove properties after constructor, you can (re-)initialize them to `undefined` if necessary.

I won't hold the PR for this though :slightly_smiling_face: